Sources/ApolloCodegenLib/ApolloSchemaDownloadConfiguration.swift

```diff
@@ -0,0 +1,90 @@
+import Foundation
+
+/// A configuration object that defines behavior for schema download.
+public struct ApolloSchemaDownloadConfiguration {
+  
+  /// How to attempt to download your schema
+  public enum DownloadMethod: Equatable {
+
+    case registry(_ settings: RegistrySettings)
+    ///   - endpointURL: The endpoint to hit to download your schema.
+    case introspection(endpointURL: URL)
+    
+    public struct RegistrySettings: Equatable {
+      public let apiKey: String
+      public let graphID: String
+      public let variant: String?
+      
+      /// Designated initializer
+      ///
+      /// - Parameters:
+      ///   - apiKey: The API key to use when retrieving your schema.
+      ///   - graphID: The identifier of the graph to fetch. Can be found in Apollo Studio.
+      ///   - variant: The variant of the graph to fetch. Defaults to "current", which will return whatever is set to the current variant.
+      public init(apiKey: String,
+                  graphID: String,
+                  variant: String = "current") {
+        self.apiKey = apiKey
+        self.graphID = graphID
+        self.variant = variant
+      }
+    }
+    
+    public static func == (lhs: DownloadMethod, rhs: DownloadMethod) -> Bool {
+      switch (lhs, rhs) {
+      case (.introspection(let lhsURL), introspection(let rhsURL)):
+        return lhsURL == rhsURL
+      case (.registry(let lhsSettings),
+            .registry(let rhsSettings)):
+        return lhsSettings == rhsSettings
+      default:
+        return false
+      }
+    }
+
+  }
+  
+  public struct HTTPHeader: Equatable, CustomDebugStringConvertible {
+    let key: String
+    let value: String
+    
+    public var debugDescription: String {
+      "\(key): \(value)"
+    }
+  }
+
+  let downloadMethod: DownloadMethod
+  let downloadTimeout: Double
+  let headers: [HTTPHeader]
+  let outputURL: URL
+
+  /// Designated Initializer
+  ///
+  /// - Parameters:
+  ///   - downloadMethod: How to download your schema.
+  ///   - downloadTimeout: The maximum time to wait before indicating that the download timed out, in seconds. Defaults to 30 seconds.
+  ///   - headers: [optional] Any additional headers to include when retrieving your schema. Defaults to nil
+  ///   - outputFolderURL: The URL of the folder in which the downloaded schema should be written
+  ///   - schemaFilename: The name, without an extension, for your schema file. Defaults to `"schema"
+  public init(using downloadMethod: DownloadMethod,
```

Comment:
For me, I would probably put the documentation above each of the actual properties on lines 56-59. Do we want to duplicate the same docs in both places? Is it better to have it on the initializer than the properties? I feel like I want to be able to see that info on the properties when I look at them also to understand what's going on.

@designatednerd What is the best way for us to handle documentation in situations like this. As far as the generated documentation, is there a preferred method for this?

Reply:
The problem with doing it above the properties and not here is that then the generated docs don't get info about the properties. It's probably a good, if very duplicative, idea to have the docs on the properties too though

Reply:
Yeah, that's what I thought. I think we should duplicate it in both places then.

Reply:
It's needed in both places; generated docs and at declaration site for Xcode. I'll duplicate them.

Reply:
I updated a few other places with documentation too, in 7fffa63f.